packages/eslint-plugin/src/rules/init-declarations.ts

```diff
@@ -0,0 +1,61 @@
+import {
+  TSESTree,
+  AST_NODE_TYPES,
+} from '@typescript-eslint/experimental-utils';
+import baseRule from 'eslint/lib/rules/init-declarations';
+import {
+  InferOptionsTypeFromRule,
+  InferMessageIdsTypeFromRule,
+  createRule,
+  deepMerge,
+} from '../util';
+
+export type Options = InferOptionsTypeFromRule<typeof baseRule>;
+export type MessageIds = InferMessageIdsTypeFromRule<typeof baseRule>;
+
+const schema = deepMerge(
+  Array.isArray(baseRule.meta.schema)
+    ? baseRule.meta.schema[0]
+    : baseRule.meta.schema,
+);
+
+export default createRule<Options, MessageIds>({
+  name: 'init-declarations',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        'require or disallow initialization in variable declarations',
+      category: 'Variables',
+      recommended: false,
+      extendsBaseRule: true,
+    },
+    schema,
+    messages: baseRule.meta.messages,
+  },
+  defaultOptions: ['always'],
+  create(context) {
+    const rules = baseRule.create(context);
+    const mode = context.options[0] || 'always';
+
+    return {
+      'VariableDeclaration:exit'(
+        node: TSESTree.VariableDeclaration,
+      ): void {
+        if (mode === 'always') {
+          if (node?.declare) {
+            return;
+          }
+          if (
+            node?.parent?.type === AST_NODE_TYPES.TSModuleBlock &&
+            node?.parent?.parent?.declare
```

Comment:
Unlike in plain javascript where you can make assumptions and write working code that does few checks, when writing typescript you have to be explicit.

In this instance, we as engineers can logically say that a TSModuleBlock can only have a parent that is TSModuleDeclaration.
However the types are setup more generally such that a TSModuleBlock has a parent that is of type Node - i.e. the types represent that any node type could be its parent.

Now for safety - typescript does not let you access a property unless it exists on every single type in a union. This is done for safety and simplicity.

So you have to do one of two things:
1) assert that `declare` exists on `node.parent.parent`.
2) assert that `node.parent.parent` is of type TSModuleDeclaration.

(1) is a lot less desired than (2), as it does not assert on the expected structure of the AST.

Reply:
this is giving error free check.

```ts
   node.parent = <TSESTree.TSModuleBlock>node?.parent;
   node.parent.parent = <TSESTree.TSModuleDeclaration>(
            node?.parent?.parent
          );
```

Reply:
should I add this ?

Reply:
that's one way to do it, but it is a type-only check. This means that if for some weird reason the node isn't a TSModuleDeclaration, so it's now made the code unsafe.

A _better_ solution would be a runtime assertion, which typescript can use to narrow the type for you:
```suggestion
            node.parent?.type === AST_NODE_TYPES.TSModuleBlock &&
            node.parent?.parent.type === AST_NODE_TYPES. TSModuleDeclaration &&
            node.parent?.parent?.declare
```